src/models/poll.ts

```diff
@@ -0,0 +1,178 @@
+        return this.pollEvent!;
+    }
+
+    public async getResponses(): Promise<Relations> {
+        // if we have already fetched the responses
+        // just return them
+        if (this.responses) {
+            return this.responses;
+        }
+        if (!this.fetchingResponsesPromise) {
+            this.fetchingResponsesPromise = this.fetchResponses();
+        }
+        await this.fetchingResponsesPromise;
+        return this.responses!;
+    }
+
+    /**
+     *
+     * @param event - event with a relation to the rootEvent
+     * @returns void
+     */
+    public onNewRelation(event: MatrixEvent): void {
+        if (M_POLL_END.matches(event.getType())) {
+            this.endEvent = event;
+            this.refilterResponsesOnEnd();
+            this.emit(PollEvent.End);
+        }
+
+        // wait for poll responses to be initialised
+        if (!this.responses) {
+            return;
+        }
+
+        const pollEndTimestamp = this.endEvent?.getTs() || Number.MAX_SAFE_INTEGER;
+        const { responseEvents } = filterResponseRelations([event], pollEndTimestamp);
+
+        if (responseEvents.length) {
+            responseEvents.forEach((event) => {
+                this.responses!.addEvent(event);
+            });
+            this.emit(PollEvent.Responses, this.responses);
+        }
+    }
+
+    private async fetchResponses(): Promise<void> {
+        this.fetchingResponsesPromise = new Promise<void>(() => {});
+
+        // we want:
+        // - stable and unstable M_POLL_RESPONSE
+        // - stable and unstable M_POLL_END
+        // so make one api call and filter by event type client side
+        const allRelations = await this.matrixClient.relations(this.roomId, this.rootEvent.getId()!, "m.reference");
+
+        // @TODO(kerrya) paging results
+
+        const responses = new Relations("m.reference", M_POLL_RESPONSE.name, this.matrixClient, [
+            M_POLL_RESPONSE.altName!,
+        ]);
+
+        const pollEndEvent = allRelations.events.find((event) => M_POLL_END.matches(event.getType()));
+        const pollCloseTimestamp = pollEndEvent?.getTs() || Number.MAX_SAFE_INTEGER;
+
+        const { responseEvents } = filterResponseRelations(allRelations.events, pollCloseTimestamp);
+
+        responseEvents.forEach((event) => {
+            responses.addEvent(event);
+        });
+
+        this.responses = responses;
+        this.endEvent = pollEndEvent;
+        if (this.endEvent) {
+            this.emit(PollEvent.End);
+        }
+        this.emit(PollEvent.Responses, this.responses);
+    }
+
+    private refilterResponsesOnEnd(): void {
+        if (!this.responses) {
+            return;
+        }
+
+        const pollEndTimestamp = this.endEvent?.getTs() || Number.MAX_SAFE_INTEGER;
+        this.responses.getRelations().forEach((event) => {
+            if (event.getTs() > pollEndTimestamp) {
+                this.responses?.removeEvent(event);
+            }
```

Comment:
For the read receipt code (defining what has been read or not) it usually prefers looking at the order in the DAG and as a last result will use the timestamp.

Could we take a similar approach here?
You have functions like `compareEventOrdering` in `EventTimelineSet` that could be good to use.

`getEventReadUpTo` is a good place to look at the logic in place that favours ordering first, and timestamp as a fallback

Reply:
I understand why this is probably more correct, but since the polls spec specifically mentions the poll end event timestamp, and comparing events by `origin_server_ts`, I think it is best to stay with strict timestamp comparison.

> Only a user's most recent vote (by `origin_server_ts`) is accepted, even if that event is invalid or redacted. Votes with timestamps after the poll has closed are ignored, as if they never happened.

> The m.poll.end's origin_server_ts determines when the poll closes exactly: if no valid end event is received, the poll is still open. If the poll is closed, only votes sent on or before that timestamp are considered, even if those votes are from before the start event. This is to handle clock drift over federation as gracefully as possible.

https://github.com/matrix-org/matrix-spec-proposals/blob/travis/msc/polls/proposals/3381-polls.md